### PR TITLE
Visual tests - Avoid using golden file tests

### PR DESCRIPTION
**Pull request by akalburgi**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by akalburgi** [comment]
WIP. @cocreature This is somehow looking complicated let me know if you feel I am using the ShakeTest setup up in an unexpected way

**Comment by akalburgi** [comment]
WIP

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 1]
```diff
@@ -89,6 +91,7 @@ data ShakeTestError
     | ExpectedVirtualResourceNote VirtualResource T.Text (Map VirtualResource T.Text)
     | ExpectedNoVirtualResourceNote VirtualResource (Map VirtualResource T.Text)
     | ExpectedNoErrors [D.FileDiagnostic]
+    | ExpectedNoMisMatch String String
```
Comment (akalburgi) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: Maybe this is a bad idea or should I move the tests which use this to this module?

File: compiler/damlc/tests/src/VisualTest.hs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import qualified Development.IDE.Core.API as API
+import qualified Development.IDE.Core.Rules.Daml as API
+import qualified Data.NameMap as NM
+import qualified Control.Monad.Reader   as Reader
+import Development.IDE.Core.API.Testing
+import System.Environment.Blank (setEnv)
+import Control.Monad.Except
+
+data ExpectedChoices = ExpectedChoices
+    { _cName :: String
```
Comment (akalburgi) [inline compiler/damlc/tests/src/VisualTest.hs]: As I am not referring the name directly within the tests, I get warning so had to use `_`

File: compiler/damlc/tests/src/VisualTest.hs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                , "        do return ()"
+                ]
+            setFilesOfInterest [foo]
+            expectedPoperties foo [TemplateProp [ExpectedChoices "Archive" True, ExpectedChoices "Delete" True] 0]
+    ]
+
+
+
+
+-- main :: IO ()
```
Comment (akalburgi) [inline compiler/damlc/tests/src/VisualTest.hs]: I feel like we should also have one of these to make sure dar reading and unpacking has not broken?

File: compiler/damlc/tests/BUILD.bazel [hunk 4]
```diff
@@ -27,15 +27,17 @@ da_haskell_test(
 # Tests for damlc visual
 da_haskell_test(
     name = "damlc-visual",
-    srcs = ["src/DamlcVisual.hs"],
+    srcs = ["src/VisualTest.hs"],
```
Comment (cocreature) [inline compiler/damlc/tests/BUILD.bazel]: I believe that change is inconsistent with the module names for other tests.

File: compiler/damlc/tests/BUILD.bazel [hunk 5]
```diff
@@ -27,15 +27,17 @@ da_haskell_test(
 # Tests for damlc visual
 da_haskell_test(
```
Comment (cocreature) [inline compiler/damlc/tests/BUILD.bazel]: I am not sure we gain much by having this be a separate test suite from the existing shake tests so I would propose that you just move it to that test suite.

File: compiler/damlc/tests/src/DamlcVisual.hs [hunk 6]
```diff
@@ -1,39 +0,0 @@
...
-
-main :: IO ()
-main = defaultMain  =<< unitTests
-
-unitTests :: IO TestTree
-unitTests = do
-    withTempFile $ \path -> do
-        darPath <- locateRunfiles (mainWorkspace </> "compiler/damlc/tests/visual-test-daml.dar")
-        dotFile <- locateRunfiles (mainWorkspace </> "compiler/damlc/tests/visual/Basic.dot")
-        return $ testGroup "making sure we do not add extra edges" [
```
Comment (cocreature) [inline compiler/damlc/tests/src/DamlcVisual.hs]: The non-golden tests still seem to be useful but are only testing the `daml-lf-reader` and not the visualize code so I would propose that we keep this, remove the golden tests and rename it to something like `//compiler/damlc/tests:daml-lf-reader`.

File: compiler/damlc/tests/src/VisualTest.hs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        Tasty.assertBool ("Shake test resulted in an error: " ++ show res) $ isRight res
+
+templateChoicesToProps :: TemplateChoices -> TemplateProp
+templateChoicesToProps tca  = TemplateProp choicesInTpl actl
+    where choicesInTpl = map (\ca -> ExpectedChoices ( DAP.renderPretty $ choiceName ca) (choiceConsuming ca)) (choiceAndActions tca)
+          actl = sum $ map (length . actions ) (choiceAndActions tca)
+
+graphTest :: LF.World -> LF.Package -> [TemplateProp] -> Either (String, String) ()
+graphTest wrld lfPkg tplProps =
+    if tplProps == map templateChoicesToProps tplPropsActual
```
Comment (cocreature) [inline compiler/damlc/tests/src/VisualTest.hs]: Does the order matter here or do we only want to test that the lists have the same elements?

File: compiler/damlc/tests/src/VisualTest.hs (hunk 3 above)
Comment (cocreature) [reply compiler/damlc/tests/src/VisualTest.hs]: Commented on this above.

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 8]
```diff
@@ -9,7 +9,9 @@
 
 -- | Testing framework for Shake API.
 module Development.IDE.Core.API.Testing
-    ( ShakeTest
+    ( ShakeTest (..)
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: I would prefer to not expose all those internals.
Instead, add a new combinator to this module something like `expectTemplateProps :: NormalizedFilePath -> [TemplateProp] -> ShakeTest ()` that takes care of all of this.

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs (hunk 1 above)
Comment (cocreature) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: `ExpectedNoMismatch` seems a bit too generator. How about `ExpectedTemplateProps [TemplateProps] [TemplateProps]` where the two lists correspond to actual and expected.

File: compiler/damlc/tests/src/VisualTest.hs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+expectedPoperties damlFilePath tplProps = do
+    ideState <- ShakeTest $ Reader.asks steService
+    mbDalf <- liftIO $ API.runAction ideState (API.getDalf damlFilePath)
+    case mbDalf of
+        Just lfPkg -> do
+            wrld <- Reader.liftIO $ API.runAction ideState (API.worldForFile damlFilePath)
+            case graphTest wrld lfPkg tplProps of
+                Right _ -> pure ()
+                Left (expected , actual) -> throwError (ExpectedNoMisMatch expected actual)
+        Nothing -> throwError (ExpectedNoErrors [])
```
Comment (cocreature) [inline compiler/damlc/tests/src/VisualTest.hs]: `ExpectedNoErrors []` isn’t right here: The list you are passing to it should be the lists of errors you got but by passing the empty lists you are indicating that you got no errors which doesn’t make sense. I would suggest that you use `expectNoErrors` and then just turn the pattern match failure here into an exception.

File: compiler/damlc/tests/src/VisualTest.hs (hunk 7 above)
Comment (akalburgi) [reply compiler/damlc/tests/src/VisualTest.hs]: No order should not matter, I will change it to Set

File: compiler/damlc/tests/src/VisualTest.hs (hunk 3 above)
Comment (akalburgi) [reply compiler/damlc/tests/src/VisualTest.hs]: renamed this to dar-reader test

File: compiler/damlc/tests/BUILD.bazel (hunk 5 above)
Comment (akalburgi) [reply compiler/damlc/tests/BUILD.bazel]: Moved all this to shake test suite

File: compiler/damlc/tests/src/DamlcVisual.hs (hunk 6 above)
Comment (akalburgi) [reply compiler/damlc/tests/src/DamlcVisual.hs]: Yup brought them back and renamed it is DarReaderTest

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs (hunk 8 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: Done

File: compiler/damlc/tests/src/VisualTest.hs (hunk 9 above)
Comment (akalburgi) [reply compiler/damlc/tests/src/VisualTest.hs]: Now returning back the expected and the actual, in this case, the actual will be empty

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 10]
```diff
@@ -10,9 +10,12 @@
 -- | Testing framework for Shake API.
 module Development.IDE.Core.API.Testing
     ( ShakeTest
+    , TemplateProp(..)
+    , ExpectedChoices(..)
     , GoToDefinitionPattern (..)
     , HoverExpectation (..)
     , D.DiagnosticSeverity(..)
+    , expectedPoperties
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]:
```suggestion
    , expectTemplateProperties
```
Matches the naming scheme of the other `expect*` functions.
I would also move it down in the export list to be next to those.

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 11]
```diff
@@ -491,6 +509,29 @@ timedSection targetDiffTime block = do
         throwError $ TimedSectionTookTooLong targetDiffTime actualDiffTime
     return value
 
+
+templateChoicesToProps :: TemplateChoices -> TemplateProp
+templateChoicesToProps tca = TemplateProp choicesInTpl (sum $ map (length . actions ) (choiceAndActions tca))
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]:
```suggestion
templateChoicesToProps tca = TemplateProp choicesInTpl (sum $ map (length . actions) (choiceAndActions tca))
```

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 12]
```diff
@@ -107,6 +116,15 @@ data ShakeTestEnv = ShakeTestEnv
     , steVirtualResourcesNotes :: TVar (Map VirtualResource T.Text)
     }
 
+data ExpectedChoices = ExpectedChoices
+    { _cName :: String
+    , _consuming :: Bool
+    } deriving (Eq, Ord, Show )
+data TemplateProp = TemplateProp
+    { _choices :: Set.Set ExpectedChoices
+    , _action :: Int
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: A comment explaining what the `Int` represents would be nice. It looks like it is the sum of the number of actions for each choice? In that case `numActions` might be a better field name. Seems like a slightly weird property to test for but we can always change or extend this later so I’m happy to keep it for now.

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 13]
```diff
@@ -491,6 +509,29 @@ timedSection targetDiffTime block = do
         throwError $ TimedSectionTookTooLong targetDiffTime actualDiffTime
     return value
 
+
+templateChoicesToProps :: TemplateChoices -> TemplateProp
+templateChoicesToProps tca = TemplateProp choicesInTpl (sum $ map (length . actions ) (choiceAndActions tca))
+    where choicesInTpl = Set.fromList $ map (\ca -> ExpectedChoices ( DAP.renderPretty $ choiceName ca) (choiceConsuming ca)) (choiceAndActions tca)
+
+graphTest :: LF.World -> LF.Package -> Set.Set TemplateProp -> Either [TemplateProp] ()
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]:
```suggestion
graphTest :: LF.World -> LF.Package -> Set.Set TemplateProp -> ShakeTest ()
```

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 14]
```diff
@@ -491,6 +509,29 @@ timedSection targetDiffTime block = do
...
+graphTest wrld lfPkg expectedProps =
+    if expectedProps == (Set.fromList $ map templateChoicesToProps tplPropsActual)
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: I would suggest that you move this to `ShakeTest`. That way you can slightly simply things by not having to pattern match on the result at the callsite.
```haskell
graphTest :: LF.World -> LF.Package -> Set.Set TemplateProp -> ShakeTest ()
graphTest wrld lfPkg expectedProps = do
    let actual = Set.fromList $ map templateChoicesToProps tplPropsActual
    unless expectedProps == actual $
        throwError $ ExpectedTemplateProps expectedProps actual

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 15]
```diff
@@ -491,6 +509,29 @@ timedSection targetDiffTime block = do
...
+        then Right ()
+        else Left $ map templateChoicesToProps tplPropsActual
+    where tplPropsActual = concatMap (moduleAndTemplates wrld) (NM.toList $ LF.packageModules lfPkg)
+
+expectedPoperties :: D.NormalizedFilePath -> Set.Set TemplateProp -> ShakeTest ()
+expectedPoperties damlFilePath expectedProps = do
+    ideState <- ShakeTest $ Reader.asks steService
+    mbDalf <- liftIO $ API.runAction ideState (API.getDalf damlFilePath)
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: I would add an `expectNoErrors` call here. That way you get usable diagnostics if it fails.
And then, I would just turn the `Nothing` case into a pattern match failure, e.g.,
`Just dalf <- pure mbDalf`

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs (hunk 12 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: Yes this `Int` is not adding any value, this is intended to check the action from a choice. I will fill this out in a following PR

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs (hunk 10 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: Yup makes sense. Done

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs (hunk 11 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: :(

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs (hunk 14 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: cool. That looks better. Done

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs (hunk 15 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: Thanks for the help with this, Using monad fail now.

**Pull request merged**